pyfar/dsp/dsp.py

```diff
@@ -312,6 +312,263 @@ def spectrogram(signal, window='hann', window_length=1024,
+            the phase. The logarithm is reversed before returning the averaged
+            signal.
+
+        The default is ``'time'``
+    phase_copy: tuple
+        Some averaging modes discard the phase in which case the phase from
+        a channel before averaging can be copied to the averaged signal. Pass
+        the channel index as a tuple with as many elements as
+        ``signal.cshape``. The default ``None`` returns a zero-phase signal.
+    weights: array like
+        array that gives channel weights for averaging the data. Must be of
+        shape ``signal.cshape``. The default is ``None``, which applies equal
+        weights to all channels.
+
+    Returns
+    --------
+    averaged_signal: Signal
+        averaged input Signal
+
+    Notes
+    -----
+    The functions :py:func:`~pyfar.dsp.linear_phase` and
+    :py:func:`~pyfar.dsp.minimum_phase` can be used to obtain non-zero phase
+    responses. This can be usefull if the average mode discards the phase and
+    the paramter `phase_copy` can not be used.
+    """
+
+    # check input
+    if not isinstance(signal, pyfar.Signal):
+        raise TypeError('Input data has to be of type: Signal')
+
+    # set weights default
+    if weights is None:
+        weights = 1/(np.prod(signal.cshape))
+    else:
+        weights = weights/np.sum(weights)
+
+    # convert data to desired domain
+    if average_mode == 'time':
+        data = signal.time.copy()
+    elif average_mode == 'complex':
+        data = signal.freq.copy()
+    elif average_mode == 'magnitude':
+        data = np.abs(signal.freq.copy())
+    elif average_mode == 'power':
+        data = np.abs(signal.freq.copy())**2
+    elif average_mode == 'log_magnitude':
+        data = 20 * np.log10(signal.freq.copy())
+
+    # apply weights
+    # NOT SURE IF THIS WORKS WITH MORE THAN FOR SIGNALS GREATER THAN 3D
+    data = data * np.transpose(np.array([weights, ]*len(data[-1])))
+
+    # average the data
+    if (average_mode == 'time' or
+            average_mode == 'complex' or average_mode == 'magnitude'):
+        data = np.sum(data, axis=-2, keepdims=True)
+    elif average_mode == 'power':
+        data = np.sum(data, axis=-2, keepdims=True)
+        data = np.sqrt(data)
+    elif average_mode == 'log_magnitude':
+        data = np.sum(data, axis=-2, keepdims=True)
+        data = 10**(data/20)
+    else:
+        raise ValueError(
+            """average_mode must be 'time', 'complex', 'magnitude', 'power' or
+            'log_magnitude'"""
+            )
+
+    # phase handling
+    if phase_copy is None:
+        pass
+    else:
+        if average_mode == 'time':
+            data_ang = signal.time.copy()
+            # NOT SURE IF THIS COPPIES PHASE_COPY INDEX IN CORRECTLY
+            data_ang = np.angle(data_ang[phase_copy, ...])
+        else:
+            data_ang = signal.freq.copy()
+            data_ang = np.angle(data_ang[phase_copy, ...])
+
+        data = data * np.exp(1j * data_ang)
```

Comment:
phase copy must work in the frequency domain. I would suggest to do after the averaged data is added to the new Signal/FrequencyData object